internal/topology/check/compatibility.go

```diff
@@ -176,10 +177,155 @@ func LocalObjectTemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace str
 		allErrs = append(allErrs,
 			field.Invalid(
 				pathPrefix.Child("ref", "apiVersion"),
-				in.Ref.APIVersion,
+				template.Ref.APIVersion,
 				"value cannot be empty",
 			),
 		)
 	}
 	return allErrs
 }
+
+// ClusterClassesAreCompatible checks the compatibility between new and old versions of a Cluster Class.
+// It checks that:
+// 1) InfrastructureCluster Templates are compatible.
+// 2) ControlPlane Templates are compatible.
+// 3) ControlPlane InfrastructureMachineTemplates are compatible.
+// 4) MachineDeploymentClasses have not been deleted and are compatible.
+func ClusterClassesAreCompatible(current, desired *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if current == nil {
+		return nil
+	}
+
+	// Validate InfrastructureClusterTemplate changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.Infrastructure, desired.Spec.Infrastructure, field.NewPath("spec", "infrastructure"))...)
+
+	// Validate control plane changes desired a compatible way.
+	allErrs = append(allErrs, LocalObjectTemplatesAreCompatible(current.Spec.ControlPlane.LocalObjectTemplate, desired.Spec.ControlPlane.LocalObjectTemplate, field.NewPath("spec", "controlplane", "localobjecttemplate"))...)
```

Comment:
Can we use the JSON paths as they are visible to the user, i.e.
* in this case localobjecttemplate should be dropped
* controlplane machineinfrastructure should be controlPlane machineInfrastructure

The current paths reflect the Go structure but that's not what a user is writing in YAML or other clients send to the server

(same in other cases)

Reply:
I think they should be up to date now.